src/soter/soter_kdf.h

```diff
@@ -57,6 +57,46 @@ soter_status_t soter_kdf(const void* key,
                          size_t context_count,
                          void* output,
                          size_t output_length);
+
+/**
+ * Computes PKCS#5 PBKDF2 HMAC-SHA-256 for a passphrase.
+ *
+ * @param [in]  passphrase          passphrase used for derivation
+ * @param [in]  passphrase_length   length of `passphrase` in bytes
+ * @param [in]  salt                additional salt for derivation, may be NULL
+ * @param [in]  salt_length         length of `salt` in bytes
+ * @param [in]  iterations          PBKDF2 iteration count
+ * @param [out] key                 output key buffer
+ * @param [in]  key_length          length of `key` in bytes
+ *
+ * This function derives a key from a passphrase using a salt and iteration
+ * count as specified in RFC 8018. It uses HMAC-SHA-256 as the hash function.
+ *
+ * The iteration count must be a positive number. The bigger it is, the slower
+ * the derivation, and the harder it gets for an attacker to perform a brute
+ * force attack with candidate passphrases. RFC 8018 suggests at least 1000.
+ * We suggest using at least 100,000. Generally, you experiment with values,
+ * use the biggest one that you can tolerate at the moment, and periodically
+ * reevaluate your decision and increase the count as machines get faster.
```

Comment:
Let's not advise this: changing the iteration count breaks compatibility. For example, if app developers decide to increase it in their "next release", they will break all the users' passwords, which they already store in the db.

Alternatively, we can add that to change the iteration count one must to regenerate they key.

Reply:
> Let's not advise this: changing the iteration count breaks compatibility.

Well... I thought it to be common sense that changing the input will result in a different output. But yeah, you’re right that the current wording can be interpreted as if the iteration count only influences the difficulty without changing the output. (Obviously, if it *did* behave that way then it would be a useless parameter.)

The derived keys and passwords should definitely not be stored in the database. If the developers would like to derive the keys themselves they should store all parameters: the salt, the iteration count, and the algorithm choice. In this case the derived key will be recoverable even if the defaults change with time.

I believe that reviewing iteration count is necessary in long-term perspective so I’d leave this advice here. However, it *is* effectively key migration so we should describe it as such for the avoidance of doubt.